File: agent/session.py
from __future__ import annotations

import json
import uuid
from pathlib import Path


def new_session_id() -> str:
    return str(uuid.uuid4())


def append_turn(sessions_dir: Path, session_id: str, messages: list[dict]) -> None:
    sessions_dir.mkdir(parents=True, exist_ok=True)
    path = sessions_dir / f"{session_id}.jsonl"
    with open(path, "a") as f:
        for msg in messages:
            f.write(json.dumps(msg, ensure_ascii=False) + "\n")


def load_session(sessions_dir: Path, session_id: str) -> list[dict]:
    path = sessions_dir / f"{session_id}.jsonl"
    if not path.exists():
        return []
    messages = []
    with open(path) as f:
        for line in f:
            line = line.strip()
            if line:
                messages.append(json.loads(line))
    return messages

File: src/agent/session.ts
import { mkdirSync, existsSync } from "fs";
import { join } from "path";
import { randomUUID } from "crypto";
import type { Message } from "../providers/base.ts";

export function newSessionId(): string {
  return randomUUID();
}

export async function appendTurn(
  sessionsDir: string,
  sessionId: string,
  messages: Message[],
): Promise<void> {
  mkdirSync(sessionsDir, { recursive: true });
  const path = join(sessionsDir, `${sessionId}.jsonl`);
  const file = Bun.file(path);
  const writer = file.writer({ flags: "a" });
  for (const msg of messages) {
    writer.write(JSON.stringify(msg, null) + "\n");
  }
  await writer.end();
}

export function loadSession(sessionsDir: string, sessionId: string): Message[] {
  const path = join(sessionsDir, `${sessionId}.jsonl`);
  if (!existsSync(path)) return [];

  const content = require("fs").readFileSync(path, "utf-8") as string;
  const messages: Message[] = [];
  for (const line of content.split("\n")) {
    const trimmed = line.trim();
    if (trimmed) {
      messages.push(JSON.parse(trimmed) as Message);
    }
  }
  return messages;
}
